### PR TITLE
Build experimental multi-arch docker image

### PR DESCRIPTION
**Pull request by Onlinehead**
Add a new Dockerfile and additional step to the uberjar workflow to build and push an experimental multi-arch Metabase container based on the Ubuntu Temurin JRE 11 image. Partially address https://github.com/metabase/metabase/issues/13119.
Now only for the `master` branch.

Some content - we almost have no hope that the Temurin team will deliver a stable Alpine-based multi-arch JRE 11 image (soon), so we have to switch to something else to unlock a possibility of running Metabase images on the arm-based machines. As the first step - we want to start to build master-based images in addition to our current alpine-based main image. 
More details about the request and difficulties in #13119 and https://github.com/adoptium/containers/issues/158.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29857)

### CODE REVIEW

File: .github/workflows/uberjar.yml [hunk 1]
```diff
@@ -100,6 +100,12 @@ jobs:
         path: bin/docker/
     - name: Move the ${{ matrix.edition }} uberjar to the context dir
       run: mv bin/docker/target/uberjar/metabase.jar bin/docker/.
+    # We need it for multi-arch build
+    - name: Set up QEMU
+      if: ${{ github.ref_name == 'master' }}
+      uses: docker/setup-qemu-action@v2
+      with:
```
Comment (paoliniluis) [inline .github/workflows/uberjar.yml:107]: I don't see this in their documentation, where did you get this flag from?

File: .github/workflows/uberjar.yml [hunk 2]
```diff
@@ -139,6 +170,15 @@ jobs:
       if: ${{ !(startsWith(github.ref_name,'master') || startsWith(github.ref_name,'backport')) && matrix.edition == 'ee' }}
       run: docker tag localhost:5000/metabase-dev:${{ steps.extract_branch.outputs.branch }}-ee metabase/metabase-dev:${{ steps.extract_branch.outputs.branch }} && docker push metabase/metabase-dev:${{ steps.extract_branch.outputs.branch }}
 
+      # Push experimental ubuntu image only for versions based on a master
+    - name: Retag and push ubuntu-based images if master (ee)
+      if: ${{ (github.ref_name == 'master') && matrix.edition == 'ee' }}
+      run: docker tag localhost:5000/metabase-dev:${{ steps.extract_branch.outputs.branch }}-ee-ubuntu metabase/metabase-enterprise-head-ubuntu:latest && docker push metabase/metabase-enterprise-head-ubuntu:latest
```
Comment (paoliniluis) [inline .github/workflows/uberjar.yml:176]: we need to create this repos right?

File: bin/docker/Dockerfile_ubuntu [hunk 3]
```diff
@@ -0,0 +1,28 @@
+FROM eclipse-temurin:11-jre-jammy as runner
```
Comment (paoliniluis) [inline bin/docker/Dockerfile_ubuntu:1]: I would just use eclipse-temurin:11-jre

File: bin/docker/Dockerfile_ubuntu [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  mkdir -p /plugins && chmod a+rwx /plugins && \
+  keytool -list -cacerts
+
+# add Metabase jar & add our run script to the image
+COPY ./metabase.jar ./run_metabase.sh /app/
+
+# expose our default runtime port
+EXPOSE 3000
+
+# if you have an H2 database that you want to initialize the new Metabase
```
Comment (paoliniluis) [inline bin/docker/Dockerfile_ubuntu:22]: I would remove these comments from the Dockerfile, I prefer to have comments in the docs

File: .github/workflows/uberjar.yml (hunk 1 above)
Comment (Onlinehead) [reply .github/workflows/uberjar.yml:107]: It is a bit not obvious, but it is required based on the Docker documentation for buildX - https://docs.docker.com/build/ci/github-actions/multi-platform/

File: .github/workflows/uberjar.yml (hunk 2 above)
Comment (Onlinehead) [reply .github/workflows/uberjar.yml:176]: That is not a separate repo, that is just another one tag. The repo is the same - metabase, but the tag has an appendix "ubuntu" now.

File: bin/docker/Dockerfile_ubuntu (hunk 3 above)
Comment (Onlinehead) [reply bin/docker/Dockerfile_ubuntu:1]: No real difference, just a bit more control over the underlayer version.

File: bin/docker/Dockerfile_ubuntu (hunk 4 above)
Comment (Onlinehead) [reply bin/docker/Dockerfile_ubuntu:22]: We have that comment in our current dockerfile for Alpine. I can remove it, but that is kind of an out of scope, I tried to save it as close to the current image as possible.

File: bin/docker/Dockerfile_ubuntu [hunk 5]
```diff
@@ -0,0 +1,28 @@
+FROM eclipse-temurin:11-jre-jammy as runner
+
+ENV FC_LANG=en-US LC_CTYPE=en_US.UTF-8
+
+# Dependencies
+RUN apt-get update && \
+  apt-get upgrade -y && \
+  apt-get install -y ca-certificates ca-certificates-java && \
```
Comment (paoliniluis) [inline bin/docker/Dockerfile_ubuntu:8]: you need fonts-dejavu as well

File: .github/workflows/uberjar.yml (hunk 1 above)
Comment (paoliniluis) [reply .github/workflows/uberjar.yml:107]: If it works, doesn't hurt my feelings

File: bin/docker/Dockerfile_ubuntu (hunk 5 above)
Comment (Onlinehead) [reply bin/docker/Dockerfile_ubuntu]: Add them.

**Review by dragonsahead — APPROVED** [review]

File: .github/workflows/uberjar.yml [hunk 6]
```diff
@@ -122,6 +128,31 @@ jobs:
     - name: Wait for Metabase to start and reach 100% health
       run: while ! curl -s 'http://localhost:3000/api/health' | grep '{"status":"ok"}'; do sleep 1; done
       timeout-minutes: 3
+
+      # Build experimental ubuntu-based images only for master
+    - name: Build ${{ matrix.edition }} Ubuntu based multi-arch container
```
Comment (ariya) [inline .github/workflows/uberjar.yml:133]: The job `containerize_test_and_push_container` is already longer and hard to decipher. Adding these steps is going to further reduce this job's readability.

Suggestion: create a new job, something like `containerize_multi_arch` or something along that line, and pack these steps into the said job instead.

Bonus: if something is wrong with the `containerize_multi_arch`, it does not hamper the progress of ``containerize_test_and_push_container` and this specific failed job can be rerun independently.

Comment (Onlinehead) [reply .github/workflows/uberjar.yml]: @ariya Good idea! Reimplemented as a separate job.

**Comment by ariya** [comment]
No matter what I did, I couldn't get the CI working on my fork, when applying this PR. It fails at the health check step:

image

@Onlinehead Do you have any example of successful output from your own test?

**Comment by paoliniluis** [comment]
@ariya what does the curl responds?

**Comment by ariya** [comment]
> @ariya what does the curl responds?

```curl: (52) Empty reply from server```

File: .github/workflows/uberjar.yml [hunk 7]
```diff
@@ -186,3 +186,75 @@ jobs:
...
+          tags: localhost:5000/metabase-dev:${{ steps.extract_branch.outputs.branch }}-${{ matrix.edition }}-ubuntu
+          no-cache: true
+          push: true
+      - name: Launch ${{ matrix.edition }} Ubuntu based container
+        run: docker run --rm -dp 3001:3000 localhost:5000/metabase-dev:${{ steps.extract_branch.outputs.branch }}-${{ matrix.edition }}-ubuntu
+        timeout-minutes: 5
+      - name: Is Docker with Ubuntu running?
+        run: docker ps
+      - name: Wait for Ubuntu-based Metabase container to start and reach 100% health
+        run: while ! curl -s 'http://localhost:3001/api/health' | grep '{"status":"ok"}'; do sleep 1; done
```
Comment (ariya) [inline .github/workflows/uberjar.yml:246]: In my test, no luck getting past this step.

**Comment by Onlinehead** [comment]
@ariya thanks for checking it. 
I re-validated it and fixed the script which prevents Metabase to start in case of a local root container (it was incompatible with ubuntu version, now it works ok with both Alpine and Ubuntu).

**Review by dragonsahead — APPROVED** [review]

**Review by ariya — APPROVED** [review]

**Comment by ariya** [comment]
@Onlinehead  Note that the push to Docker Hub still failed (credential issue?), see https://github.com/metabase/metabase/actions/runs/4691289365/jobs/8315943495#step:15:14.

image

**Comment by ariya** [comment]
> Note that the push to Docker Hub still failed (credential issue?), see https://github.com/metabase/metabase/actions/runs/4691289365/jobs/8315943495#step:15:14.

Nevermind, looks like this was fixed already in PR #30080.

**Pull request merged**